File: LeetCode/Study Plan/Practice Problems/Medium/Longest Palindrome By Concatenating Two Letter Words/Go/longest_palindrome_by_concatenating_two_letter_words.go
package main

import "fmt"

func longest_palindrome_by_concatenating_two_letter_words(words []string) int {
	mp := make(map[string]int)
	for i := 0; i < len(words); i++ {
		mp[words[i]]++
	}
	odd := 0
	l := 0
	val := ""
	for k, v := range mp {
		if k[0] == k[1] && v%2 == 1 {
			if v > odd {
				val = k
				odd = v
			}
		}
	}
	l += odd * 2
	mp[val] = 0
	for i := 0; i < len(words); i++ {
		k := words[i]
		if mp[k] > 0 {
			if k[0] == k[1] {
				if mp[k]%2 == 0 {
					l += mp[k] * 2
				} else {
					l += (mp[k] - 1) * 2
				}
				mp[k] = 0
			} else {
				l += (min(mp[k], mp[string(k[1])+string(k[0])]) * 4)
				mp[k] = 0
				mp[string(k[1])+string(k[0])] = 0
			}
		}
	}
	return l
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func main() {
	fmt.Println(longest_palindrome_by_concatenating_two_letter_words([]string{"lc", "cl", "gg"}))
}

File: LeetCode/Medium/Longest Palindrome By Concatenating Two Letter Words/Cpp/longest_palindrome_by_concatenating_two_letter_words.cc
#include <iostream>
#include <map>
#include <vector>

typedef std::vector<std::string> vs;
typedef std::map<std::string, int> msi;

int longest_palindrome_by_concatenating_two_letter_words(vs &v) {
    msi cnt;
    std::string s = "";
    int odd = 0, ans = 0;
    for (int i=0;i<v.size(); i++) cnt[v[i]]++;
    for (auto it=cnt.begin(); it!=cnt.end(); it++){
        std::string word = it->first;
        if (word[0] == word[1]) {
            if (cnt[word]%2 && cnt[word] > odd) {
                s = word;
                odd = cnt[word];
            }
        }
    }
    ans += odd*2;
    cnt[s] = 0;
    //handle words with same characters
    for (int i=0;i<v.size(); i++) {
        std::string word = v[i];
        int c = cnt[word];
        if (word[0] == word[1] && c > 0) {
            if (c % 2)  ans += (c-1)*2;
            else ans += c * 2;
            cnt[word] = 0;
        }
    }
    //handle words with different characters
    for (int i=0;i<v.size();i++){
        std::string word = v[i];
        int c = cnt[word];
        if (c > 0) {
            std::string rev = word;
            reverse(rev.begin(), rev.end());
            ans += std::min(cnt[word], cnt[rev])*4;
            cnt[word] = 0;
            cnt[rev] = 0;
        }
    }
    return ans;
}

int main() {
    vs v{"nn","nn","hg","gn","nn","hh","gh","nn","nh","nh"};
    std::cout<<longest_palindrome_by_concatenating_two_letter_words(v)<<"\n";
} 
